### PR TITLE
[next] Add handling for app-paths-manifest

### PR DESCRIPTION
**Pull request by ijjk**
This adds handling for the `app` outputs and adds initial tests to ensure it is working as expected. 

### Related Issues

x-ref: https://github.com/vercel/next.js/pull/38420

#### Tests

- [ ] The code changed/added as part of this PR has been covered with tests
- [ ] All tests pass locally with `yarn test-unit`

#### Code Review

- [ ] This PR has a concise title and thorough description useful to a reviewer
- [ ] Issue from task tracker has a link to this PR

### CODE REVIEW

File: packages/next/src/index.ts [hunk 1]
```diff
@@ -1071,10 +1071,18 @@ export const build: BuildV2 = async ({
       'pages'
     );
 
-    const pages = await getServerlessPages({
+    const appPathRoutesManifest = await readFile(
```
Comment (TooTallNate) [inline packages/next/src/index.ts]: Could use `readJSON()` here

Comment (ijjk) [reply packages/next/src/index.ts]: Updated in https://github.com/vercel/vercel/pull/8098/commits/c2301a954b21a6bd73517a74d65efd2a05051cd1

File: packages/next/src/index.ts [hunk 2]
```diff
@@ -2595,18 +2609,39 @@ async function getServerlessPages(params: {
...
+    params.appPathRoutesManifest
+      ? glob('**/page.js', path.join(params.pagesDir, '../app'))
+      : Promise.resolve({}),
     getMiddlewareManifest(params.entryPath, params.outputDirectory),
   ]);
 
+  const normalizedAppPaths: typeof appPaths = {};
+
+  if (params.appPathRoutesManifest) {
+    for (const entry of Object.keys(params.appPathRoutesManifest || {})) {
```
Comment (styfle) [inline packages/next/src/index.ts]: Nit pick: use `Object.entries()` here to get key and value at the same time

File: packages/next/src/index.ts [hunk 3]
```diff
@@ -2595,18 +2609,39 @@ async function getServerlessPages(params: {
...
+        normalizedAppPaths[normalizedPath] = appPaths[globPath];
+      }
+    }
+  }
+
   // Edge Functions do not consider as Serverless Functions
   for (const edgeFunctionFile of Object.keys(
     middlewareManifest?.functions ?? {}
   )) {
+    delete normalizedAppPaths[edgeFunctionFile.slice(1) + '.js'];
```
Comment (styfle) [inline packages/next/src/index.ts]: Nit pick: use a variable like 
```js
const edgePath = edgeFunctionFile.slice(1) + '.js';
delete normalizedAppPaths[edgePath];
delete pages[edgePath];
```

File: packages/next/src/server-build.ts [hunk 4]
```diff
@@ -128,11 +133,21 @@ export async function serverBuild({
   prerenderManifest: NextPrerenderedRoutes;
   requiredServerFilesManifest: NextRequiredServerFilesManifest;
 }): Promise<BuildResult> {
+  Object.assign(lambdaPages, lambdaAppPaths);
```
Comment (styfle) [inline packages/next/src/server-build.ts]: Is it safe to mutate the `lambdaPages` here? What if the calling function uses it after calling this function?

Comment (ijjk) [reply packages/next/src/server-build.ts]: Should be safe as the builder returns the result of the server build directly. Updated to create a separate object though.

File: packages/next/src/index.ts (hunk 3 above)
Comment (ijjk) [reply packages/next/src/index.ts]: Updated in https://github.com/vercel/vercel/pull/8098/commits/8953047b143e9134b30a4b9f12acaabada795af0

File: packages/next/src/index.ts (hunk 2 above)
Comment (ijjk) [reply packages/next/src/index.ts]: Updated in https://github.com/vercel/vercel/pull/8098/commits/8953047b143e9134b30a4b9f12acaabada795af0

File: packages/next/src/index.ts [hunk 5]
```diff
@@ -2606,18 +2620,39 @@ async function getServerlessPages(params: {
...
+      ? glob('**/page.js', path.join(params.pagesDir, '../app'))
+      : Promise.resolve({}),
     getMiddlewareManifest(params.entryPath, params.outputDirectory),
   ]);
 
+  const normalizedAppPaths: typeof appPaths = {};
+
+  if (params.appPathRoutesManifest) {
+    for (const [entry, normalizedEntry] of Object.entries(
+      params.appPathRoutesManifest || {}
```
Comment (styfle) [inline packages/next/src/index.ts]: You can remove the `if` on line 2635 or remove the `|| {}` on line 2637

Comment (ijjk) [reply packages/next/src/index.ts]: Updated in https://github.com/vercel/vercel/pull/8098/commits/5a04b53b8eb34e8ea4e07cfb63ff307c81b9df69

**Review by styfle — APPROVED** [review]

**Pull request merged**